server/src/main/java/io/crate/types/DataTypes.java

```diff
@@ -365,50 +363,27 @@ public static DataType<?> upcast(DataType<?> type) {
      * @param upcast    if true, then use the widest possible compatible numeric type
      *                  e.g. a list of integers produces array(long)
      */
-    public static DataType<?> valueFromList(List<?> value, boolean upcast) {
+    public static DataType<?> typeFromList(List<?> value, boolean upcast) {
         DataType<?> highest = DataTypes.UNDEFINED;
+        int idx = 0;
         for (Object o : value) {
             if (o == null) {
                 continue;
             }
             DataType<?> current = guessType(o);
-            // JSON libraries tend to optimize things like [ 0.0, 1.2 ] to [ 0, 1.2 ]; so we allow mixed types
-            // in such cases.
-            if (!current.equals(highest) && !safeConversionPossible(current, highest)) {
-                throw new IllegalArgumentException(
-                    "Mixed dataTypes inside a list are not supported. Found " + highest + " and " + current);
-            }
-            if (current.precedes(highest)) {
+            if (idx == 0) {
                 highest = current;
+            } else {
+                highest = DataTypes.merge(highest, current);
```

Comment:
minor: can we simplify this by removing `idx`? I think `DataTypes.merge` can handle `idx==0` case.